[Bian Yuqi] iP

Jacky142857 opened the pull request:
# Duke
> "The key is not to prioritize what is on your schedule, but to schedule your priorities." -- Stephen Covey

DukePro frees your mind of having to remember things you need to do. It's,
- text based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [ here ](https://github.com/Jacky142857/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Jacky142857 committed:
Level-1

Jacky142857 committed:
Level-2

Jacky142857 committed:
Level-3

Jacky142857 committed:
added event, deadline and todo subclass

Jacky142857 committed:
first/partial test case done

Jacky142857 committed:
added more detailed tests

Jacky142857 committed:
some exceptions handled

Jacky142857 committed:
delete function

Jacky142857 committed:
Add TaskList class

Jacky142857 committed:
Add save function

Jacky142857 committed:
Revert "Add save function"

This reverts commit b3d7954681a548b096afd056342d78e28f86d85e.

Jacky142857 committed:
Add save function

Jacky142857 committed:
Add load function

Jacky142857 committed:
Add date recognization function if input time is a date

Jacky142857 committed:
Merge branch 'branch-Level-7'

Jacky142857 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java

Resolution: Keep both constructors

Jacky142857 committed:
Add Storage class and Parser class

Jacky142857 committed:
Extract out Ui class, add more methods to Parser class

Jacky142857 committed:
Complete extraction of Parser, Ui, Storage class

Jacky142857 committed:
Organize class into packages

Classes are organized into duke package, with 3 packages under it.

duke package contains helper, exception and tasks package.

Jacky142857 committed:
Add Junit Test with Gradle

Test methods in Todo and Parser class

Jacky142857 committed:
Modify the files in accordance to coding standards

Jacky142857 committed:
Add JavaDoc to classes and methods

Jacky142857 committed:
Implement find function to search for a keyword

Jacky142857 committed:
Merge branch 'branch-A-JavaDoc'

Jacky142857 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/tasks/Task.java

Jacky142857 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/helper/Parser.java

Resolution: Keep the code that is applicable for find function

Jacky142857 committed:
Add gradle to automate build and test

Jacky142857 committed:
Add more JavaDoc comments

cwq2326 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,213 @@
+        String command = strArr[0];
+        if (!(command.equals("bye") || command.equals("list") || command.equals("mark") || command.equals("todo")
+                || command.equals("deadline") || command.equals("event") || command.equals("delete") ||
+                command.equals("find"))) {
```
Can include all these commands in an enum and call create a method to check if all these commands fall under the category to avoid complicated expressions.

cwq2326 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,213 @@
+            if (hasKeyword(strArr, "/by")) {
+                String title;
+                String[] splitArr = input.split("/by", 2);
+                try{
+                    title = splitArr[0].substring("deadline".length() + 1).trim();
+                }
+                catch (StringIndexOutOfBoundsException e){
+                    throw new DukeEmptyArgumentException();
+                }
```
Missing spacing after try and close bracket for the catch statement, and catch should be directly after curly braces
```suggestion
                try {
                    title = splitArr[0].substring("deadline".length() + 1).trim();
                } catch (StringIndexOutOfBoundsException e) {
                    throw new DukeEmptyArgumentException();
                }
```

cwq2326 commented on src/main/java/duke/helper/TaskList.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+
+    public Task get(int i){
```
Maybe can rename this to getTask 
```suggestion
    public Task getTask(int i) {
```

cwq2326 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+import duke.helper.TaskList;
+import duke.helper.Ui;
+
+/**
+ *
+ */
```
Missing header comment

cwq2326 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,213 @@
+                return new Event(strArr[2], ld, isDone);
+            }
+            return new Event(strArr[2], strArr[3], isDone);
+        default:
+            return null;
```
Good that you included the default branch

cwq2326 commented on src/main/java/duke/helper/Parser.java:
```diff
@@ -0,0 +1,213 @@
+            if (hasKeyword(strArr, "/at")) {
+                String title;
+                String[] splitArr = input.split("/at", 2);
+                try {
```
Maybe can put the try to try all of ur code and catch the exception so dont have so many trys and catch (Just a suggestion, not sure if it will be more complicated to do this)
try {
// all your different commands
} catch (error1) {
} catch (error2) {...}

cwq2326 commented on src/main/java/duke/exception/DukeInvalidFileException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exception;
+
+import duke.exception.DukeException;
+
+public class DukeInvalidFileException extends DukeException {
+    @Override
+    public String toString() {
+        return "Something wrong with the file";
+    }
+}
```
I like how you separate the exceptions into different ones so that it is more clear what type of exception has occured.

cwq2326 reviewed (commented):
Overall good code quality and code looks clean.
However, there are some minor indentation and spacing issues.

Jacky142857 committed:
Implement level-10 GUI for Duke

Jacky142857 committed:
Merge branch 'branch-Level-10'

fantablack commented on src/main/java/duke/exception/DukeEmptyArgumentException.java:
```diff
@@ -0,0 +1,10 @@
+
+import duke.exception.DukeException;
+
+public class DukeEmptyArgumentException extends DukeException {
+    @Override
+    public String toString() {
+        return "arguments shouldn't be empty";
+    }
+}
```
I like how you made use of inheritance to represent different types of DukeExceptions rather than just using different Strings to represent different exceptions

fantablack commented on src/main/java/duke/helper/Storage.java:
```diff
@@ -0,0 +1,74 @@
+
+
+    public Storage(String filePath) {
+        String[] pathArr = filePath.split("/");
+        String directoryPath = "/" + pathArr[0];
```
Its great that you managed to managed to make this work using Strings to represent paths, but I suggest looking into the newer Java Path library if you have the time 😊

fantablack commented on src/main/java/duke/helper/Storage.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            file = new File(System.getProperty("user.dir") + "/" + filePath);
+            if (!file.isFile()) {
+                File dir;
+                dir = new File(System.getProperty("user.dir") + directoryPath);
+                dir.mkdir();
+                file.createNewFile();
```
I like this simple and straightforward way of creating directories and file(s). I would like to suggest using the newer Java NIO Paths and Files library to do something similar such as:
```
 Files.createDirectories(directoryPath);
 Files.createFile(filePath);
```

fantablack reviewed (commented):
Overall, I found your code easy to read without being needlessly complicated and jam-packed with unnecessary OOP. Great job! I came across some newer APIs such as Files and Path NIOs in my research that I would love to share with you as they are supposedly faster. However, your usage of the older API works just fine too. Keep it up 👍

Jacky142857 committed:
Add welcome and goodbye message to GUI

The welcome and goodbye message of GUI is missing.

Users will not find this application interactive as it shows nothing at
the start and close abruptly at the end.

Adding a greeting message at the start makes the user feel more welcomed
and adding a goodbye message and a 1.5 second delay before closing will
make the experience of user more natural.

Jacky142857 committed:
Add assertions to Parser.java

Assertions are made to make sure the logic of storing tasks into file is
implemented correctly.

Each line of the file should begin with one of the 'E', 'T', or 'D'. And
assertion also checks that each line has correct number of components
separated by delimeter '|'.

Jacky142857 committed:
Improve code quality of the code

code in Parser.java are reordered to remove nesting, variables 'ld' are
renamed to 'localDate' to improve readability. Greet method are
extracted out from MainWindow.java to follow SLAP.

Jacky142857 committed:
Merge pull request #1 from Jacky142857/branch-A-Assertions

Add assertions to Parser.java

Jacky142857 committed:
Merge branch 'master' into branch-A-CodeQuality

Jacky142857 committed:
Remove useless code in Ui.java since we already have GUI.

Jacky142857 committed:
Merge pull request #2 from Jacky142857/branch-A-CodeQuality

Branch a code quality

Jacky142857 committed:
Implement C-Archive function

Sometimes user might want to start a new task list while still keeping
a copy of their previous task list, they can use archive function to
achieve this.

User can archive their task list by typing 'archive' and add their
archived list back to the task list they are working with by typing
'unarchive'.

Jacky142857 committed:
Fix coding style and coding standard issues.

Jacky142857 committed:
Set theme jekyll-theme-cayman

Jacky142857 committed:
Fix bug on empty argument with find keyword

Jacky142857 committed:
Fix bug on mark command when there's no argument provided

Jacky142857 committed:
Fix the spacing of output message

Jacky142857 committed:
Update user guide

Jacky142857 committed:
Add Ui.png

Jacky142857 committed:
Update Ui.png

Jacky142857 committed:
Update README.md

Jacky142857 committed:
Update README.md
